```javascript
// js에는 동기함수와 비동기 함수가 있다, syncronous는  '동시에 일어나는' 이라는 뜻으로 어떤 작업의 응답과 다음 작업의 요청이 동시에 일어나는 것을 뜻한다

// js의 비동기함수를 호출하면 해당 함수의 응답을 기다리지 않고 바로 다음 작업을 실행한다, 즉 js에서 비동기라는것은 여러가지 작업을 동시에 실행 할 수 있는 것이고 응답을 기다리지 않는다는 것을 뜻한다, 동기함수는 작업의 응답과 다음 작업이 동시에 발생하므로 순차적으로 일이 진행되게 된다, 즉 동기함수의 종료, 반환을 기다린다

// 비동기함수를 호출하는 것은 성능면에서 좋지만 비동기함수가 뭔가를 반환한 다음에 그것을 이용해서 순차적으로 일을 진행해야하는 경우에는 원하는 대로 작업을 처리할 수 가 없다

// 이때 비동기함수가 순차적으로 실행되도록 하는 것을 비동기 처리라고 한다

// js 에서 비동기 처리는 콜백함수를 이용한다

// 함수의 인자로 전달되어 함수 내부에서 호출되는 함수를 콜백함수라고 한다, 함수가 인자로 전달될 수 있는 것은 js가 일급함수를 쓰기 때문이다 first class function

  

// callback함수를 이용한 비동기 처리

// 비동기 함수가 할일을 끝내고 제일 마지막에 다음으로 순차적으로 할 일을 하는 함수(콜백함수)를 호출한다

  

function orderPizza(callback) {
  console.log('order pizza');
  setTimeout(() => {
    const pizza = 'real pizza';
    console.log('pizza making is done');
    callback(pizza);
  }, 2000);
}

  
function eatPizza(pizza) {
  console.log(`Eat ${pizza}`);
}

  
console.log('play game');
orderPizza(eatPizza);
console.log('do the dishes');
```

```javascript
// promise를 이용한 비동기 처리

// promise는 미래에 얻을 데이터가 있다고 가정하고 그 데이터로 할일을 작성할수있게 해주는 객체이다

// 미래에 실제로 데이터를 얻는다, 데이터를 전달한다

// resolve함수에 데이터를 전달하며 오류가 발생했을경우에는 reject로 에러 메세지를 전달한다

// 비동기함수가 promise객체를 반환하도록 작성하여 비동기처리를 한다

  
let pizza;
function orderPizza() {
  console.log('order Pizza');
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('pizza making is done');
      resolve('real pizza');
    }, 2000);
  });
}

  
function deliverPizza(box) {
  console.log('deliver start');
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('delivery is done');
      reject(box + ' in box');
    }, 2000);
  });
}

function eatPizza(pizza) {
  console.log(`Eat ${pizza}`);
}

  

// promise 객체로 부터 데이터를 가져와 다음 할일을 계속 작성하려면 promise객체의 메소드인 then을 이용한다

// 그리고 에러를 다루기 위해 promise객체의 catch 메소드를 이용한다

// catch함수로 에러를 전역적으로 다룰 수 있다

// then의 두번째 매개변수로 에러를 다루는 함수를 전달할 수 있지만 그렇게 하면 성공적으로 실행되었다고 판단하고 다음 then을 이어서 실행한다

  
console.log('play game');
orderPizza()
  .then(deliverPizza)
  .then(eatPizza)
  .catch((error) => {
    console.log(`Error: ${error}`);
  });
console.log('do the dishes');
```

```javascript
// async, await은 promise를 이용하는 방법중 하나이며 기존의 then과 문법이 다를뿐이며 promise객체를 다룬다

// 어떤 함수든 async함수로 만들수 있으며 async함수 내에서만 await을 쓸 수 있다

// await은 promise를 반환하는 함수를 호출할때만 사용할 수 있다

// await은 promise가 데이터를 가져오기를 기다린다 (즉 비동기 함수를 동기함수처럼 사용한다)

// 콜백이나 then과 달리 기존의 동기함수처럼 어떤 값을 반환하는 함수처럼 비동기함수를 사용할 수 있다

// try/catch로 에러를 다뤄야 한다

  

function orderPizza() {
  console.log('order Pizza');
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('pizza making is done');
      resolve('real pizza');
      //reject('pizza making went something wrong');
    }, 2000);
  });
}

  

function deliverPizza(pizza) {
  console.log('deliver start');
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('delivery is done');
      resolve(pizza + ' in box');
    }, 2000);
  });
}

  

function eatPizza(pizza) {
  console.log(`Eat ${pizza}`);
}

  

// catch를 해줘야 reject로 전달된 에러메세지를 확인할 수 있다!!

// 비동기 함수를 만들고 그 함수 내에서 할일들을 비동기함수가 순차적으로 진행한다(비동기 함수가 await키워드를 통해 동기적으로 실행된다)

// 해당 함수 자체는 비동기함수이므로 async함수 자체의 응답을 기다리지 않으며 해당함수내에서 순차적으로 할일을 알아서 끝낸다

// await은 promise객체를 반환하는 함수를 호출할때만 사용 가능하다!

async function callDomino() {
  try {
    const pizza = await orderPizza();
    const box = await deliverPizza(pizza);
    eatPizza(box);
  } catch (error) {
    console.log(error);
  }
}

  
console.log('play game');
callDomino(); // 비동기함수이므로 함수가 종료될때까지 기다리지 않고 다음 코드를 실행한다
console.log('do the dishes');
```